pkg/server/server.go

```diff
@@ -0,0 +1,78 @@
+// package server is an implementation of the remote-exec server.
+package server
+
+import (
+	"context"
+	"crypto/x509"
+	"fmt"
+
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/peer"
+
+	"github.com/mlaterman/remote-exec/pkg/auth"
+	pb "github.com/mlaterman/remote-exec/pkg/proto"
+)
+
+type ExecServer struct {
+	pb.UnimplementedExecutorServer
+}
+
+func New() *ExecServer {
+	return &ExecServer{}
+}
+
+func (e *ExecServer) checkCert(ctx context.Context) error {
+	cert, err := getClientCert(ctx)
+	if err != nil {
+		return err
+	}
+	name := "" // TODO get name value
+	return auth.Authz(cert, name)
+}
+
+func getClientCert(ctx context.Context) (*x509.Certificate, error) {
+	p, ok := peer.FromContext(ctx)
+	if !ok {
+		return nil, fmt.Errorf("call has no peer information")
+	}
+	tlsInfo, ok := p.AuthInfo.(credentials.TLSInfo)
+	if !ok {
+		return nil, fmt.Errorf("unable to gather tls info")
+	}
+	certs := tlsInfo.State.PeerCertificates
+	if len(certs) == 0 {
+		return nil, fmt.Errorf("no peer certificates found")
+	}
+	return certs[0], nil
+}
+
+func (e *ExecServer) Start(ctx context.Context, startProc *pb.StartProcess) (*pb.ProcessID, error) {
+	err := e.checkCert(ctx)
+	if err != nil {
+		return nil, err
+	}
+	return nil, fmt.Errorf("start not implemented")
+}
+func (e *ExecServer) Stop(ctx context.Context, id *pb.ProcessID) (*pb.StopResponse, error) {
+	err := e.checkCert(ctx)
+	if err != nil {
+		return nil, err
+	}
+	return nil, fmt.Errorf("stop not implemented")
```

Comment:
You probably want to return a GRPC error here.

https://pkg.go.dev/google.golang.org/grpc/status#Error
https://grpc.github.io/grpc/core/md_doc_statuscodes.html

Reply:
changed, also changed the cert check errors to return unauthenticated error codes